### PR TITLE
[ENG-3898][ENG-3899]ResourceDetail PATCH and DELETE

### PR DESCRIPTION
**Pull request by jwalz**
## Purpose
Implement PATCH and DELETE methods on the ResourceDetail endpoint

## Changes
* Update ResourceDetailPermission to allow PATCH and DELETE
* Implement `ResourceSerializer.update()` to handle PATCHing
* Implement `ResourceDetail.perform_destroy()` to handle DELETEing
* Add EnumFieldMemberError Exception class
  * Raise in EnumField when an unsupported value is encountered
  * Catch in `ResourceDetail.patch()` to populate the `source` field
* Tests

## Ticket
https://openscience.atlassian.net/browse/ENG-3898
https://openscience.atlassian.net/browse/ENG-3899

### CODE REVIEW

File: api/resources/serializers.py [hunk 1]
```diff
@@ -72,6 +78,68 @@ def create(self, validated_data):
         try:
             root_outcome = Outcome.objects.for_registration(primary_registration, create=True)
         except NoPIDError:
-            raise Conflict('Cannot add Resources to a Registration that does not have a DOI')
+            raise Conflict('Cannot add Resrouces to a Registration without a DOI')
```
Comment (brianjgeiger) [inline api/resources/serializers.py:81]:
```suggestion
            raise Conflict('Cannot add Resources to a Registration without a DOI')
```

Comment (mfraezz) [inline api/resources/serializers.py:81]: This line should be reverted

File: api/resources/serializers.py [hunk 2]
```diff
@@ -72,6 +78,68 @@ def create(self, validated_data):
...
 
         return OutcomeArtifact.objects.create(outcome=root_outcome)
+
+    def update(self, instance, validated_data):
+        updated_title = validated_data.get('title')
+        if updated_title is not None:
+            instance.title = updated_title
```
Comment (mfraezz) [inline api/resources/serializers.py:88]: Minor/Clarity: didn't we decide that `title` might be auto-populated but not user-inputted?

File: api_tests/resources/views/test_resource_detail.py [hunk 3]
```diff
@@ -40,7 +72,8 @@ def test_status_code__unapproved__non_contributor(self, app, user_role):
             registration_state=RegStates.INITIAL, user_role=user_role
         )
         resp = app.get(make_api_url(test_artifact), auth=test_auth, expect_errors=True)
-        assert resp.status_code == 403 if test_auth else 401
+        expected_status_code = 403 if test_auth else 401
+        assert resp.status_code == expected_status_code
```
Comment (brianjgeiger) [inline api_tests/resources/views/test_resource_detail.py:76]: Out of curiosity, why the change? Just better readability?

File: api/resources/views.py [hunk 4]
```diff
@@ -77,3 +77,6 @@ def get_object(self):
 
     def get_permissions_proxy(self):
         return Guid.load(self.get_object().primary_resource_guid).referent
+
+    def perform_destroy(self, instance):
+        instance.delete(api_request=self.request)
```
Comment (mfraezz) [inline api/resources/views.py:82]: Should this be `instance.deleted = timezone.now()`?

File: api_tests/resources/views/test_resource_detail.py [hunk 5]
```diff
@@ -131,28 +166,346 @@ def test_anonymized_data(self, app):
...
+
+        app.patch_json_api(make_api_url(test_artifact), payload, auth=test_auth, expect_errors=True)
+        test_artifact.refresh_from_db()
+
+        assert test_artifact.title == original_title
+        assert test_artifact.description == original_description
+        assert test_artifact.artifact_type == ArtifactTypes.UNDEFINED
+        assert test_artifact.identifier == original_identifier
+        assert test_artifact.identifier.value == original_identifier.value
+        assert not test_artifact.finalized
```
Comment (brianjgeiger) [inline api_tests/resources/views/test_resource_detail.py:394]: So `finalized` has to be PATCHed by itself?

File: api_tests/resources/views/test_resource_detail.py [hunk 6]
```diff
@@ -131,28 +166,346 @@ def test_anonymized_data(self, app):
...
+
+    @pytest.mark.parametrize('registration_state', [RegStates.ACCEPTED, RegStates.EMBARGO])
+    def test_status_code__admin(self, app, registration_state):
+        test_artifact, test_auth, _ = configure_test_preconditions(
+            registration_state=registration_state, user_role=UserRoles.ADMIN_USER
+        )
+        resp = app.delete_json_api(make_api_url(test_artifact), auth=test_auth, expect_errors=True)
+        assert resp.status_code == 204
+
+    @pytest.mark.parametrize('user_role', UserRoles.excluding(UserRoles.ADMIN_USER))
```
Comment (brianjgeiger) [inline api_tests/resources/views/test_resource_detail.py:422]: So Write users can add and edit, but only Admins can delete? That seems odd.

File: osf/models/outcomes.py [hunk 7]
```diff
@@ -88,5 +88,6 @@ def log_artifact_change(self, action, artifact, api_request, **log_params):
         self.primary_osf_resource.add_log(
             action=nodelog_action,
             params=nodelog_params,
+            auth=None,  # Grabbed from request
```
Comment (mfraezz) [inline osf/models/outcomes.py:91]: Is this necessary?

**Review by mfraezz — CHANGES_REQUESTED** [review]
A couple questions, but generally looks good. :octocat:

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
A few questions, maybe one typo.

File: api/resources/serializers.py (hunk 1 above)
Comment (brianjgeiger) [reply api/resources/serializers.py:81]: NVM, I like Matt's suggestion better to revert.

File: api/resources/views.py (hunk 4 above)
Comment (jwalz) [reply api/resources/views.py:82]: See the custom delete logic from https://github.com/CenterForOpenScience/osf.io/pull/9989 (and the tests)

File: osf/models/outcomes.py (hunk 7 above)
Comment (jwalz) [reply osf/models/outcomes.py:91]: required positional argument 🤦

File: api_tests/resources/views/test_resource_detail.py (hunk 5 above)
Comment (jwalz) [reply api_tests/resources/views/test_resource_detail.py:394]: no, this was just the clearest way to show that a failed PATCH reverts all changes

File: api_tests/resources/views/test_resource_detail.py (hunk 3 above)
Comment (jwalz) [reply api_tests/resources/views/test_resource_detail.py:76]: Better error results on the tests. Having a condition on the RHS just leads to an `assert False` instead of showing the actual vs expected code. Need to go and fix this on the DELETEPermissions tests, too

File: api/resources/serializers.py (hunk 1 above)
Comment (jwalz) [reply api/resources/serializers.py:81]: Thanks! Missed this one

File: api_tests/resources/views/test_resource_detail.py (hunk 6 above)
Comment (jwalz) [reply api_tests/resources/views/test_resource_detail.py:422]: Wrote this before we agreed that WRITE users should be able to POST, in which case WRITE users should absolutely be able to DELETE. Good catch, thanks!

**Review by brianjgeiger — APPROVED** [review]

File: api/resources/serializers.py [hunk 8]
```diff
@@ -75,3 +79,75 @@ def create(self, validated_data):
...
         return OutcomeArtifact.objects.create(outcome=root_outcome)
+
+    def update(self, instance, validated_data):
+        updated_artifact_type = validated_data.get('artifact_type')
+        if updated_artifact_type is not None and updated_artifact_type != instance.artifact_type:
+            if updated_artifact_type == ArtifactTypes.UNDEFINED:
+                raise JSONAPIException(
+                    detail=(
+                        f'Resource with id [{instance._id}] currently has a resource_type of '
+                        f'"{instance.artifact_type}", cannot retrun resource_type to "undefined".'
```
Comment (brianjgeiger) [inline api/resources/serializers.py:90]:
```suggestion
                        f'"{instance.artifact_type}", cannot return resource_type to "undefined".'
```

File: api/resources/serializers.py [hunk 9]
```diff
@@ -75,3 +79,75 @@ def create(self, validated_data):
...
+            return
+
+        source = '/data/attributes/'
+        if len(error_sources) == 1:  # Be more specific if possible
+            source += error_sources[0]
+
+        raise Conflict(
+            detail=(
+                f'Cannot PATCH `finalized: true` for Resource with id [{instance._id}] '
+                f'until the following  required fields are populated {error_sources}.'
```
Comment (brianjgeiger) [inline api/resources/serializers.py:148]:
```suggestion
                f'until the following required fields are populated {error_sources}.'
```

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Just a couple of tiny things. Otherwise looks good.

**Review by brianjgeiger — APPROVED** [review]
